app/src/lib/git/squash.ts

```diff
@@ -0,0 +1,69 @@
+import * as FSE from 'fs-extra'
+import { getCommits, revRange } from '.'
+import { Commit, CommitOneLine } from '../../models/commit'
+import { Repository } from '../../models/repository'
+import { getTempFilePath } from '../file-system'
+import { rebaseInteractive, RebaseResult } from './rebase'
+
+/**
+ * Squashes provided commits by calling interactive rebase
+ *
+ * @param toSquash - commits to squash onto another commit
+ * @param squashOnto  - commit to squash the `toSquash` commits onto
+ * @param lastRetainedCommitRef -  sha of commit before all commits in squash operation
+ * @param commitSummary - summary of new commit made by squash
+ * @param commitDescription - description of new commit made by squash
+ */
+export async function squash(
+  repository: Repository,
+  toSquash: ReadonlyArray<CommitOneLine>,
+  squashOnto: CommitOneLine,
+  lastRetainedCommitRef: string,
+  commitSummary: string,
+  commitDescription: string
+): Promise<RebaseResult> {
+  const commits: Commit[] = [
+    ...(await getCommits(repository, revRange(lastRetainedCommitRef, 'HEAD'))),
+  ]
+
+  let todoOutput = ''
+  commits.forEach(c => {
+    if (toSquash.map(sq => sq.sha).includes(c.sha)) {
+      return
+    }
+
+    if (c.sha === squashOnto.sha) {
+      todoOutput += `pick ${c.sha} ${c.summary}\n`
+      toSquash.forEach(async sq => {
+        todoOutput += `squash ${sq.sha} ${sq.summary}\n`
+      })
+      return
+    }
+
+    todoOutput += `pick ${c.sha} ${c.summary}\n`
+  })
+
+  const todoPath = await getTempFilePath('squashTodo')
+  await FSE.writeFile(todoPath, todoOutput)
```

Comment:
Nothing to worry about in this PR (and probably ever 😂 ), but if the list of commits was too long, having the whole content of the `squashTodo` in memory before writing it to a file could kill the app.

As an alternative, we could write a line at a time directly instead of building the string.

Again, not suggesting you have to do it, I'm just thinking out loud 😆

Reply:
:D That was my original solution.. but async troubles.. But honestly didn't look into it too long. Probably easiest to just move it from a foreach to a regular for loop to make async not be mean.

Reply:
Updated to for loop/appendFile. May not have needed `for loop`... I was using `writeFile`.. when I needed to use `appendFile` originally 🤦, However, the `for loop` is also significantly more efficient which would be good under the circumstances of crazy numerous commits and this prevents having to convert commits into a modifiable array and in my next pr I realized I need to traverse the commits in reverse order so added that here and the for loop allows me to do that with out calling reverse.. which again is much more efficient. 

Also.. noticed I had unnecessarily had logic to write the pick lines twice... so refactored that too.

Basically the guts of the squash method are completely refactored.. :P